core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/V1WorkspaceMigrationMonitor.scala

```diff
@@ -140,28 +155,117 @@ object V1WorkspaceMigrationMonitor
     migrations
       .filter { m => m.workspaceId === workspace.workspaceIdAsUUID && m.started.isDefined && m.finished.isEmpty }
       .length
       .result
       .map(_ > 0)
 
+
   final def schedule(workspace: Workspace): WriteAction[Unit] =
     DBIO.seq(migrations.map(_.workspaceId) += workspace.workspaceIdAsUUID)
 
+
+  final def claimAndConfigureNewGoogleProject(migrationAttempt: V1WorkspaceMigrationAttempt,
+                                              workspaceService: WorkspaceService,
+                                              workspace: Workspace,
+                                              billingProject: RawlsBillingProject
+                                             ): IO[(GoogleProjectId, GoogleProjectNumber, WriteAction[Unit])] =
+    IO.fromFuture {
+      IO {
+        for {
+          (googleProjectId, googleProjectNumber) <- workspaceService.setupGoogleProject(
+            billingProject,
+            workspace.currentBillingAccountOnGoogleProject.getOrElse(
+              throw new RawlsException(s"""No billing account for workspace '${workspace.workspaceId}'""")
+            ),
+            workspace.workspaceId,
+            workspace.toWorkspaceName
+          )
+        } yield (googleProjectId, googleProjectNumber, DBIO.seq(
+          migrations.filter(_.id === migrationAttempt.id)
+            .map(r => (r.newGoogleProjectId, r.newGoogleProjectNumber, r.newGoogleProjectConfigured))
+            .update((googleProjectId.value.some, googleProjectNumber.value.some, Timestamp.valueOf(LocalDateTime.now).some))
+        ))
+      }
+    }
+
+
+  final def createTempBucket(attempt: V1WorkspaceMigrationAttempt,
+                             workspace: Workspace,
+                             googleStorageService: GoogleStorageService[IO]
+                            ): IO[(GcsBucketName, WriteAction[Unit])] = {
+    val tmpBucketName = GcsBucketName(
+      "terra-workspace-migration-" + UUID.randomUUID.toString.replace("-", "")
+    )
+
+    for {
+      googleProject <- attempt.newGoogleProjectId match {
+        case Some(projectId) => IO.pure(GoogleProject(projectId.value))
+        case None            => failNoGoogleProject(attempt)
+      }
+
+      _ <- createBucketInSameRegion(
+        GcsBucketName(workspace.bucketName),
+        googleProject,
+        tmpBucketName,
+        googleStorageService
+      )
+
+    } yield (tmpBucketName, DBIO.seq(
+      migrations
+        .filter(_.id === attempt.id)
+        .map(r => (r.tmpBucket, r.tmpBucketCreated))
+        .update((tmpBucketName.value.some, Timestamp.valueOf(LocalDateTime.now).some))
+    ))
+  }
+
+
+  final def createFinalBucket(attempt: V1WorkspaceMigrationAttempt,
+                              workspace: Workspace,
+                              googleStorageService: GoogleStorageService[IO]
+                            ): IO[(GcsBucketName, WriteAction[Unit])] =
+    for {
+      googleProject <- attempt.newGoogleProjectId match {
```

Comment:
It might be nice to have whatever is calling `createFinalBucket` checking for the existence of `googleProject` and `tmpBucket` so that `createFinalBucket` doesn't have to worry about where it sits in the migration process

Reply:
Interesting idea. I was thinking this would be called after the original bucket had been copied and deleted. There's got to be some ordering involved but it perhaps it could be looser than I thought.